### ISSUE TITLE
attempt to subtract with overflow error 

### ISSUE THREAD

**Issue by goulart-paul** [rwl/amd_order#3]
The following example causes a panic with error 'attempt to subtract with overflow'.

```rust
fn main() {

    let n: usize = 35;

    let a_p = vec![
        0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 17, 20, 24, 
        26, 29, 33, 35, 38, 42, 45, 47, 51, 54, 56, 61, 64, 66, 70, 74, 77];

    let a_i = vec![
         0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 0, 15, 1, 15, 16, 
         2, 15, 16, 17, 3, 18, 4, 18, 19, 5, 18, 19, 20, 6, 21, 7, 21, 22, 
         8, 21, 22, 23, 2, 11, 24, 1, 25, 0, 9, 14, 26, 5, 12, 27, 4, 28, 
         3, 9, 10, 14, 29, 8, 13, 30, 7, 31, 6, 10, 14, 32, 11, 12, 
         13, 33, 9, 10, 34];

    let mut p_inv = vec![0; 77];
    let control = amd::Control::default();
    let mut a = [[""; 77]; 77];

    amd::control(&control);

    // This causes a panic
    let (p, p_inv, info) = amd::order::<usize>(n, &a_p, &a_i, &control).unwrap();

    amd::info(&info);
}

```

The issue appears at this line:  https://github.com/rwl/amd_order/blob/258df5819cb1fbbd4cb1d504595e4ab57140625e/src/amd_2.rs#L382

I tried to find a smaller example but this was the best I could do.   I couldn't put my finger on what it is that causes this issue - it doesn't happen for all matrices and is not obviously related to the dimension.

**Comment by rwl** [comment]
Good catch. Originally, I used signed `i32` integers like the C version. Perhaps I should've added a feature to support `i64` as this would've been closer to the SuiteSparse approach, but I was trying to learn about generics in Rust:

https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/AMD/Include/amd_internal.h#L158

```rust
#[cfg(not(feature = "i64"))]
pub(crate) type Int = i32;

#[cfg(feature = "i64")]
pub(crate) type Int = i64;
```

When I get time, I'll include your example as a test. Perhaps, some sort of fuzz testing against random matrices can be added to catch any more bugs like this.

**Comment by goulart-paul** [comment]
@rwl We have been exercising this code quite hard within our interior point optimization solver ([link](https://github.com/oxfordcontrol/Clarabel.rs)), and have not had any problems other than the one here.   

I implemented a symmetric factorisation method within the solver using your AMD crate for the pre-factorisation permutation (see [this bit](https://github.com/oxfordcontrol/Clarabel.rs/blob/main/src/qdldl/qdldl.rs)).   You will see that I ended up just using `usize` everywhere for the indexing because otherwise the code becomes a mess of casts and `unwrap`s.  

We are testing it against several thousand example problems (some quite weird) over a wide range of dimensions.  We also compare results against a Julia implementation of the same code, which is using the SuiteSparse AMD implementation underneath, with identical results.  I am reasonably confident that this is working correctly and is quite fast.

Sincere thanks for putting this together.
